// solr-faceted-search-react/src/api/solr-query.js
const rangeFacetToQueryFilter = (field) => {

	const filters = field.value || [];
	if (filters.length < 2) {
		return null;
	}

	return encodeURIComponent(`${field.field}:[${filters[0]} TO ${filters[1]}]`);
};

const dateRangeFacetToQueryFilter = (field) => {
const filters = (field.value !== undefined && field.value !== null) ? field.value.toString().split(" ") : [];
	if (filters.length < 2) {
		return null;
	}
return encodeURIComponent(`${field.field}:[${filters[0]} TO ${filters[1]}]`);
};

const listFacetFieldToQueryFilter = (field) => {
	const filters = field.value || [];
	if (filters.length === 0) {
		return null;
	}

	const filterQ = filters.map((f) => `"${f}"`).join(" OR ");
	return encodeURIComponent(`${field.field}:(${filterQ})`);
};

const textFieldToQueryFilter = (field) => {
	if(!field.value || field.value.length === 0) {
		return null;
	}

	return encodeURIComponent(field.field === "*" ? field.value : `${field.field}:${field.value}`);
};

const advancedSearchFieldToQueryFilter = (field) => {
	if(!field.value || field.value.length === 0) {
		return null;
	}
	//change the regex
	var fieldVals = field.value.split(";");
	console.log(fieldVals);
	return null;
};


const fieldToQueryFilter = (field) => {
	if (field.label === "Advanced Entry (field:value,...)")
	{
		//do the value parse here
		if (field.value !== undefined)
		{
			console.log("HERE: " + field.value);
		}
		return advancedSearchFieldToQueryFilter(field) ;
	}
	else if (field.type === "text") {
		return textFieldToQueryFilter(field);
	} else if (field.type === "list-facet") {
		return listFacetFieldToQueryFilter(field);
	}else if (field.type === "date-range-facet") {
		return dateRangeFacetToQueryFilter(field);
	} else if (field.type.indexOf("range") > -1) {
		return rangeFacetToQueryFilter(field);
	}
	return null;
};

const buildQuery = (fields) => fields
	.map(fieldToQueryFilter)
	.filter((queryFilter) => queryFilter !== null)
	.map((queryFilter) => `fq=${queryFilter}`)
	.join("&");

const facetFields = (fields) => fields
	.filter((field) => field.type === "list-facet" || field.type === "range-facet" || field.type === "date-range-facet")
	.map((field) => `facet.field=${encodeURIComponent(field.field)}`)
	.join("&");


const facetSorts = (fields) => fields
	.filter((field) => field.facetSort)
	.map((field) => `f.${encodeURIComponent(field.field)}.facet.sort=${field.facetSort}`)
	.join("&");

const buildSort = (sortFields) => sortFields
	.filter((sortField) => sortField.value)
	.map((sortField) => encodeURIComponent(`${sortField.field} ${sortField.value}`))
	.join(",");

const buildFormat = (format) => Object.keys(format)
	.map((key) => `${key}=${encodeURIComponent(format[key])}`)
	.join("&");

const solrQuery = (query, format = {wt: "json"}, noAttachments = true) => {
	const {
			searchFields,
			sortFields,
			rows,
			start,
			facetLimit,
			facetSort,
			pageStrategy,
			cursorMark,
			idField
		} = query;

	const filters = (query.filters || []).map((filter) => ({...filter, type: filter.type || "text"}));
	const queryParams = buildQuery(searchFields.concat(filters));
	const addAttachment = (noAttachments) ? " AND subject_s:*": "";
	const facetFieldParam = facetFields(searchFields);
	const facetSortParams = facetSorts(searchFields);
	const facetLimitParam = `facet.limit=${facetLimit || -1}`;
	const facetSortParam = `facet.sort=${facetSort || "index"}`;
	const cursorMarkParam = pageStrategy === "cursor" ? `cursorMark=${encodeURIComponent(cursorMark || "*")}` : "";
	const idSort = pageStrategy === "cursor" ? [{field: idField, value: "asc"}] : [];

	const sortParam = buildSort(sortFields.concat(idSort));
	
	var stack = new Error().stack;
  console.log("PRINTING CALL STACK");
  console.log( stack );

return `q=*:*${encodeURIComponent(addAttachment)}` +
		 `&${queryParams.length > 0 ? queryParams : ""}` +
		`${sortParam.length > 0 ? `&sort=${sortParam}` : ""}` +
		`${facetFieldParam.length > 0 ? `&${facetFieldParam}` : ""}` +
		`${facetSortParams.length > 0 ? `&${facetSortParams}` : ""}` +
		//need to & the date range here
		`&rows=${rows}` +
		`&${facetLimitParam}` +
		`&${facetSortParam}` +
		`&${cursorMarkParam}` +
		(start === null ? "" : `&start=${start}`) +
		"&facet=on" +
		`&${buildFormat(format)}`;
};

export default solrQuery;


export {
	rangeFacetToQueryFilter,
	listFacetFieldToQueryFilter,
	textFieldToQueryFilter,
	fieldToQueryFilter,
	buildQuery,
	facetFields,
	facetSorts,
	buildSort,
	solrQuery
};
